sql/schemachanger/scplan/internal/rules: rework dep rules

ajwerner opened the pull request:
The dep rules helpers were brittle. Instead, make it easier to construct rel
queries directly.

Release note: None

ajwerner committed:
sql/schemachanger/scplan/internal/rules: rework dep rules

The dep rules helpers were brittle. Instead, make it easier to construct rel
queries directly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82256)
<!-- Reviewable:end -->

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -358,37 +383,45 @@ func init() {
+				joinOnIndexID(from, to),
+				targetStatusEq(fromTarget, toTarget, scpb.ToAbsent),
+				currentStatusEq(fromNode, toNode, scpb.Status_ABSENT),
+				rel.Filter("secondaryIndexPartialIsNotBeingDropped", from)(func(
```
Kebab-case all the rel things?

ajwerner replied:
I made the filter functions camelCase. I'm starting to dislike the kebabs. If I were typing this language I would not be pleased with kabobs I don't think. Sometimes I think about what an interpreter would look like. I'm going to keep iterating around these parts, happy to make further changes later.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/helpers.go:
```diff
@@ -32,93 +30,43 @@ func targetNodeVars(el rel.Var) (element, target, node rel.Var) {
-	registerDepRule(d.ruleName, d.edgeKind, fromNode, toNode, screl.MustQuery(c...))
+func joinOnColumnID(a, b rel.Var) rel.Clause {
+	return rel.And(
+		join(a, b, screl.DescID, "desc-id"),
```
suggestion: `relation-id`? also applies to `joinOnIndexID` and others

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -22,319 +22,344 @@ import (
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type((*scpb.TemporaryIndex)(nil)),
+				to.Type((*scpb.PrimaryIndex)(nil), (*scpb.SecondaryIndex)(nil)),
```
I wish there was a way to make these type unions pop out more systematically somehow. It's why I originally newlined each subtype, I found that it helped readability. I think this is fine though.

ajwerner replied:
Done.

postamar reviewed (approved):
LGTM, I like it and it's simpler.

ajwerner commented:
bors r+

ajwerner committed:
sql/schemachanger/rel: ensure that chosen indexes overlap

Before this change, an index which had it conditions met could be used even if
the input query did not constrain the index at all. We never want this to be
the case for queries we use; we should always be able to constrain the database
at least a little. This rule caught a bug that made logic tests slow. I think
it'll be very handy.

Note that as a special case, an index which indexes not columns permits any
scan. This is useful for testing the logic of the library.

Release note: None

ajwerner commented:
@postamar I added a new first commit to this which, in retrospect, is an important one. There was a buglet where we'd say that an index can be used even if it matches none of the attributes. This lead to the possibility of writing rules which would not be index accelerated. The hope is to always index accelerate rule evaluation. The immediately caught the bug which lead to the change making logic tests slow. PTAL before I bors it.

ajwerner commented:
I'm going to bors it.

bors r+
